src/librustc_mir/interpret/place.rs

```diff
@@ -231,33 +235,11 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
     ) -> EvalResult<'tcx, MPlaceTy<'tcx>> {
         let pointee_type = val.layout.ty.builtin_deref(true).unwrap().ty;
         let layout = self.layout_of(pointee_type)?;
-        let mplace = match self.tcx.struct_tail(pointee_type).sty {
-            // Matching on the type is okay here, because we used `struct_tail` to get to
-            // the "core" of what makes this unsized.
-            ty::Dynamic(..) => {
-                let (ptr, vtable) = val.to_scalar_dyn_trait()?;
-                MemPlace {
-                    ptr,
-                    align: layout.align,
-                    extra: PlaceExtra::Vtable(vtable),
-                }
-            }
-            ty::Str | ty::Slice(_) => {
-                let (ptr, len) = val.to_scalar_slice(self)?;
-                MemPlace {
-                    ptr,
-                    align: layout.align,
-                    extra: PlaceExtra::Length(len),
-                }
-            }
-            _ => {
-                assert!(!layout.is_unsized(), "Unhandled unsized type {:?}", pointee_type);
-                MemPlace {
-                    ptr: val.to_scalar()?,
-                    align: layout.align,
-                    extra: PlaceExtra::None,
-                }
-            }
+        let mplace = if layout.is_unsized() {
```

Comment:
won't this get us into trouble with `extern type`s? They are unsized but have no `extra`

Reply:
@eddyb says they technically have an extra of `()`. But you are right that this might still not work because their extra is not pointer-sized.

The old code here assumed that `Dynamic`, `Str` and `Slice` are the only possible unsized tails.

What could a testcase look like?

Reply:
```rust
#![feature(extern_types)]

extern {
    type Foo;
}

fn main() {
    let x: &Foo = unsafe { &*(16 as *const Foo) };
    let y: &Foo = &*x;
}
```

might already trigger it

Reply:
Fixed.